### PR TITLE
rust noob's first commit

### PR DESCRIPTION
**Pull request by tqhoughton**
added some defaults in case you get a non json response back

### CODE REVIEW

File: api/src/lib.rs [hunk 1]
```diff
@@ -84,7 +84,15 @@ pub async fn submit_request(input: HttpRequest) -> Result<Value, Box<dyn Error>>
...
+    if !res_type.starts_with("application/json") {
+        // I couldn't figure out how to safely throw an error so I'm just returning this for now
+        println!("expected application/json, got {}", res_type);
+        return Ok(json!({"msg": "not JSON!"}));
+    }
+
     let res_str = res.text().await?;
-    let res_json = serde_json::from_str(&res_str).unwrap();
+    let res_json = serde_json::from_str(&res_str).unwrap_or_default();
     Ok(res_json)
```
Comment (tqhoughton) [inline api/src/lib.rs:97]: one thing I am wondering, do we want to stick with only JSON being the valid return type for now or do we want to bake in support for non-json responses like HTML or text/binary?

fine if we go with the former approach but just wanted to make sure it was in line with your vision

File: api/src/lib.rs [hunk 2]
```diff
@@ -84,7 +84,15 @@ pub async fn submit_request(input: HttpRequest) -> Result<Value, Box<dyn Error>>
         HttpMethod::HEAD => todo!(),
     };
 
+    let res_type = res.headers().get("content-type").unwrap().to_str().unwrap();
+
+    if !res_type.starts_with("application/json") {
+        // I couldn't figure out how to safely throw an error so I'm just returning this for now
+        println!("expected application/json, got {}", res_type);
+        return Ok(json!({"msg": "not JSON!"}));
```
Comment (tqhoughton) [inline api/src/lib.rs:92]: this is 100% me being stupid, spent like an hour trying to figure out how to safely throw an error here and gave up, decided to just return a normal JSON value for now so it doesn't break the UI

File: gui/src/main.rs
Comment (tqhoughton) [inline gui/src/main.rs:1]: thought we might want a separate tab in the UI for history

**Comment by tqhoughton** [comment]
Overall the project looks sweet, it's crazy how little code is needed to get things started up :)

File: api/src/lib.rs (hunk 1 above)
Comment (mblydenburgh) [reply api/src/lib.rs:97]: I wanted to bake in support for other response types, I just defaulted to json to make something work for now. I think that once we get header support in then we can use the response type header to have a `match` statement in order to call the right deserialization method before returning to the ui.

I think that would necessitate the creation of something like a Response wrapper so the `submit_request` could still maintain a single output type? Open to some thoughts on to best handle that though.

**Comment by mblydenburgh** [comment]
> Overall the project looks sweet, it's crazy how little code is needed to get things started up :)

Yea I was pretty happy how easy it was to get going with the ui library, and cargo workspaces make making multi-crate projects really easy. I wonder when we are going to hit a limit with the gui lib. If you are curious on all its widgets they have a nice demo site with references to the source - https://www.egui.rs/#demo

File: api/src/lib.rs (hunk 2 above)
Comment (mblydenburgh) [reply api/src/lib.rs:92]: I think I ran into an issue with trying to throw an error too, relating to `Box<dyn Error>` not implementing Sync or Send or something like that I think? Obv have some refreshing to do there hehe

**Comment by mblydenburgh** [comment]
I wonder if it would make sense to create a new repo that is part of an organization? I can't seem to see this branch or checkout from my local? 

Screenshot 2023-10-12 at 8 05 21 AM

**Review by mblydenburgh — APPROVED** [review]
Achievement unlocked: First Rust PR

**Comment by tqhoughton** [comment]
> I wonder if it would make sense to create a new repo that is part of an organization? I can't seem to see this branch or checkout from my local?
> 
> Screenshot 2023-10-12 at 8 05 21 AM

That's because I created this branch in a fork of your project I think, you should be able to merge it in and then it will show up in your project I think? In the future I will make PRs where the source branch is your own project, this was me trying to get things working without access last night

**Pull request merged**